lib/src/utils.dart

```diff
@@ -43,3 +43,53 @@ YamlWarningCallback yamlWarningCallback = (message, [SourceSpan? span]) {
   if (span != null) message = span.message(message);
   print(message);
 };
+
+// The following utility functions are copied from the string_scanner package.
+//
+// See https://en.wikipedia.org/wiki/UTF-16#Code_points_from_U+010000_to_U+10FFFF
+// for documentation on how UTF-16 encoding works and definitions of various
+// related terms.
+
+/// The inclusive lower bound of Unicode's supplementary plane.
+const _supplementaryPlaneLowerBound = 0x10000;
+
+/// The inclusive upper bound of Unicode's supplementary plane.
+const _supplementaryPlaneUpperBound = 0x10FFFF;
+
+/// The inclusive lower bound of the UTF-16 high surrogate block.
+const _highSurrogateLowerBound = 0xD800;
+
+/// The inclusive lower bound of the UTF-16 low surrogate block.
+const _lowSurrogateLowerBound = 0xDC00;
+
+/// The number of low bits in each code unit of a surrogate pair that goes into
+/// determining which code point it encodes.
+const _surrogateBits = 10;
+
+/// A bit mask that covers the lower [_surrogateBits] of a code point, which can
+/// be used to extract the value of a surrogate or the low surrogate value of a
+/// code unit.
+const _surrogateValueMask = (1 << _surrogateBits) - 1;
+
+/// Returns whether [codePoint] is in the Unicode supplementary plane, and thus
+/// must be represented as a surrogate pair in UTF-16.
+bool inSupplementaryPlane(int codePoint) =>
+    codePoint >= _supplementaryPlaneLowerBound &&
+    codePoint <= _supplementaryPlaneUpperBound;
+
+/// Returns whether [codeUnit] is a UTF-16 high surrogate.
+bool isHighSurrogate(int codeUnit) =>
+    (codeUnit & ~_surrogateValueMask) == _highSurrogateLowerBound;
```

Comment:
Another option is:
```dart
 => codeUnit ^ _highSurrogateLowerBound < 0x400;
```

Reply:
If keeping the implementation with string_scanner is not needed, how about just use the implementation in [Flutter's TextPainter.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/text_painter.dart#L1269-L1285), i.e.

```dart
bool isHighSurrogate(int value) => value & 0xFC00 == 0xD800;
bool isLowSurrogate(int value) => value & 0xFC00 == 0xDC00;
```
?

Reply:
I was thinking the `0x400` would be a better immediate for fx ARM instructions, but it seems their immediates are not 12-bit, but 8-bit data + 4-bit shift, which all of 0x400, 0xFC00, 0xD800 and 0xDC00 fits in, so there should be no difference.

For compiling to Intel machine code, maybe this approoach would be better:
```dart
bool isHighSurrogate(int value) => value >> 10 == 0x36;
bool isLowSurrogate(int value) => value >> 10 == 0x37;
```
which can get implemented with one-byte immediates instead of a 32-bit ones (16-bit instructions are slow).
It should still just be two instructions on ARM.

(One can argue that this is worrying too much about small optimizations when we're not even sure the function will be inilined, and the real big saving would come from doing vector operations instead of checking individual characters. But I do worry about small stuff. :sunglasses:)